### ISSUE TITLE
Absent PoX anchors should be "forkable"

### ISSUE THREAD

**Issue by kantai** [stacks-network/stacks-blockchain#1805]
There are a few cases where Stacks blocks should be processable in multiple PoX forks. For example, if a reward cycle begins where nobody Stacked, then all of the block commits of that reward cycle would be _burns_, regardless of whether or not the anchor block has already been processed by the node. However, the node cannot _know_ that this is the case until it has processed that anchor block. This would result in a set of sortitions that would select the same sortition winners in two different PoX forks. It is _vitally_ important that those sortition winners be valid in either PoX fork. If they are not valid in either PoX fork, that would mean that _any_ missing PoX anchor block essentially stalls the blockchain (because the discovery of that anchor block would always invalidate all subsequent blocks).

There are a handful of implementation details and Clarity features that prevent blocks from being valid in multiple PoX forks:

1. The `write_consensus_bytes` implementation for back pointers uses the referred-to MARFTrieID as the consensus bytes for the back pointer. This could instead use the root hash of the referred-to trie. See: https://github.com/blockstack/stacks-blockchain/blob/cb75c8f9a555aff85eafc068909d9d573941775d/src/chainstate/stacks/index/node.rs#L311

2. The `BLOCK_HEIGHT_TO_HASH_MAPPING_KEY` and `BLOCK_HASH_TO_HEIGHT_MAPPING_KEY` key/values stored in each MARF trie. These store the parent "MarfTrieId" in each trie. Because the MarfTrieId is a function of the consensus hash and the Stacks block hash, this mixes the PoX fork identifier into the root hash. These key/value pairs are necessary for a bunch of implementation reasons-- they are how we evaluate `(at-block ...)`, compute the ancestor hash skip list, among other things. However, these probably do _not_ need to be committed to. If we could find a way to track these without storing them in the Stacks block commitment MARF, that would be very helpful.

3. Right now: exposing the `index-block-hash` in Clarity. Eventually: exposing other PoX reward cycle information.

**Comment by jcnelson** [comment]
Regarding (1), I think that, by itself, this change should be fine.  But, please see (2) and (3) below, since they may influence (1).

Regarding (2), could we do the following?
* Change the mapping to store the parent trie root hash instead of the MarfTrieId?  Then, we can still make use of the `BLOCK_HEIGHT_TO_HASH_MAPPING_KEY` and `BLOCK_HASH_TO_HEIGHT_MAPPING_KEY` key/value pairs; it's just that now they would not be dependent on any PoX state.
* Have the MARF commit to all other non-PoX information that would have been committed to by the Stacks index block hash (which is what we currently use for the MarfTrieId)?  A trivial way to do this would be to insert one extra key/value pair into the MARF that mapped the string "__BLOCK_HEADER_INFO" to the canonical serialization of the Stacks block header, with a state root hash of all 0's.

3.  I'm on the fence about exposing PoX state to Clarity.  On the one hand, I don't want to hide any useful information from developers.  But on the other hand, I don't know if the existence of PoX state in the VM can cause a block that would otherwise be valid in multiple PoX forks to now be invalid.  For example, if we expose the consensus hash to the Clarity VM, would this re-introduce the problem?

**Comment by kantai** [comment]
> Change the mapping to store the parent trie root hash instead of the MarfTrieId? Then, we can still make use of the BLOCK_HEIGHT_TO_HASH_MAPPING_KEY and BLOCK_HASH_TO_HEIGHT_MAPPING_KEY key/value pairs; it's just that now they would not be dependent on any PoX state.

Maybe -- the problem is that we use those mappings to get pointers to the trie. If we then have a way to go from trie root hash to marf-trie-id, then we should be able to do this.

> Have the MARF commit to all other non-PoX information that would have been committed to by the Stacks index block hash (which is what we currently use for the MarfTrieId)? A trivial way to do this would be to insert one extra key/value pair into the MARF that mapped the string "__BLOCK_HEADER_INFO" to the canonical serialization of the Stacks block header, with a state root hash of all 0's.

Yep, that should be doable.

> Regarding (3), I'm on the fence about exposing PoX state to Clarity. On the one hand, I don't want to hide any useful information from developers. But on the other hand, I don't know if the existence of PoX state in the VM can cause a block that would otherwise be valid in multiple PoX forks to now be invalid. For example, if we expose the consensus hash to the Clarity VM, would this re-introduce the problem?

Yep, if we exposed the consensus hash, it would re-introduce the problem -- I could write a contract that just exposed one public function for issuing a "consensus-binding" transaction, storing the most recent consensus hash into a data-var. At the start of every block, I issue that transaction. In the event of a PoX fork, any block that included that transaction would now be invalid.

**Comment by jcnelson** [comment]
> Maybe -- the problem is that we use those mappings to get pointers to the trie. If we then have a way to go from trie root hash to marf-trie-id, then we should be able to do this.

I think this is could just be a sqlite table that doesn't get committed to by the trie root hash.  All this table would do is map the (PoX-less) global trie identifier back onto the (PoX-specific) local trie identifier.  But, I'm leery about any scheme that does not force the trie to commit to the values of `BLOCK_HEIGHT_TO_HASH_MAPPING_KEY` and `BLOCK_HASH_TO_HEIGHT_MAPPING_KEY`, since I think (a) light clients are going to need to authenticate this information using only block headers (and this is trivial if these keys are just committed to already by the block headers' state root hashes), and (b) as a general design principal, I think Clarity should only ever read state that is either authenticated by the state root, or is deterministically derived in a PoX-free manner from authenticated state (such as contract metadata).

> Yep, if we exposed the consensus hash, it would re-introduce the problem -- I could write a contract that just exposed one public function for issuing a "consensus-binding" transaction, storing the most recent consensus hash into a data-var. At the start of every block, I issue that transaction. In the event of a PoX fork, any block that included that transaction would now be invalid.

Sounds good.  I'll go ahead and remove the `consensus-hash` global variable I introduced in my PoX lockup PR.

**Comment by jcnelson** [comment]
@kantai I'm still a little foggy about the precise reason why PoX state is not allowed to be represented in the Clarity VM.  We don't want to expose the consensus hash, because that could render a block that needs to be valid in two PoX forks invalid in at least one of them.  Is the fundamental reason for this because it allows the same block to evaluate to two different state-roots due to the transactions having a PoX-specific interpretation?  I'm asking because I'm trying to reason about how lazy Stacks token unlocks in PoX will affect block evaluation.  I think this concern doesn't apply here, because PoX lock state is the same in all PoX forks (and thus a block that tries to lock or unlock tokens will have the same interpretation in all PoX forks).  Is that reasoning correct?

**Comment by kantai** [comment]
> Is the fundamental reason for this because it allows the same block to evaluate to two different state-roots due to the transactions having a PoX-specific interpretation?

Yep, exactly.

>  I'm asking because I'm trying to reason about how lazy Stacks token unlocks in PoX will affect block evaluation. I think this concern doesn't apply here, because PoX lock state is the same in all PoX forks (and thus a block that tries to lock or unlock tokens will have the same interpretation in all PoX forks). Is that reasoning correct?

Yeah, that's correct.

**Comment by diwakergupta** [comment]
Note from discussion on 8/17: doesn't necessarily block PoX on Krypton, but would be a mainnet blocker.

**Comment by jcnelson** [comment]
@kantai Do you know the status of this issue?

**Comment by kantai** [comment]
Yep, this is still an open issue

**Comment by kantai** [comment]
Circling back on this, I think there's two potential ways to address this issue:

## 1. Continue to uniquely identify Stacks block tries with a function of `BlockHeaderHash` and `ConsensusHash`.

Most of the code _outside_ of the MARF implementation would continue to use `StacksBlockId` struct as currently implemented to index into the MARF. _However_, the MARF would need to use a _different_ identifier internally. This would impact the following:

1. **The root hash skip-list calculation.** This would need to use the `BlockHeaderHash` instead.
2. **Back-pointer consensus bytes in MARF proofs.** Again, rather than the `StacksBlockId`, it would use `BlockHeaderHash`.
3. **Back-pointer traversal.** Currently, MARF tries store a key/value entry in each block that resolves block height to a trie identifier. This would need to either use the `BlockHeaderHash` to perform the traversal (by implementing some method of getting a `StacksBlockId` from a `BlockHeaderHash` while indexed into a MARF "fork", _or_ resurrecting something like the fork table struct. This would be far and away the most painful part of this change.

## 2. Revert to uniquely identify Stacks block tries with a function of `BlockHeaderHash` and `BurnchainHeaderHash`.

At the surface, this seems like it's not possible (or if it is, it's suprising). But, I believe that this should be possible, even in the presence of PoX forks. This is because a winning `BlockHeaderHash/BurnchainHeaderHash` will always refer to a single possible MARF trie _even_ if a PoX fork occurs, however, the block data would need to be revalidated (because one of its parents may not have been elected). Doing this revalidation would likely require _dropping_ invalidated trie (or otherwise marking as invalid) during a PoX reorg.


From my best understanding at the moment, the difference between these two approaches is roughly that (1) implicitly enforces the revalidation of blocks by continuing to externally index by ConsensusHash/BlockHeaderHash and internally resolving backpointers with a ConsensusHash/BlockHeaderHash lookup data structure, whereas (2) does not require the same indirection because it would be explicitly dropping Stacks block data associated with a PoX reorg.

I'm currently more inclined towards option (2), because I believe it would lead to less indirection in the normal/expected path, wouldn't require introducing a new data structure (which is a source of some implementation uncertainty), and it cleans out data that we should be dropping anyways. I'd love to hear thoughts from others on these options, though, and also if you can think of any other strategies.

**Comment by jcnelson** [comment]
Popping up a level, I want to reconsider this:

> For example, if a reward cycle begins where nobody Stacked, then all of the block commits of that reward cycle would be burns, regardless of whether or not the anchor block has already been processed by the node. However, the node cannot know that this is the case until it has processed that anchor block. This would result in a set of sortitions that would select the same sortition winners in two different PoX forks. It is _vitally_ important that those sortition winners be valid in either PoX fork.

In each reward cycle, there are effectively two sets of forks:  the forks created by miners who have the anchor block (call these "PoX-descendant forks"), and the forks created by miners who do not (call these "nephew forks").  Right now, the danger is that a miner can mine a hidden anchor block, build on it privately, and release the anchor block at an arbitrarily far point in the future.  Today, this would invalidate all other forks produced -- i.e. all nephew forks created by miners who burned because they believed the anchor block did not exist would get blown away.  This is obviously bad -- it's not prohibitively expensive to pull this attack off, and it can be used by a financially-motivated attacker to blackmail users, miners, and exchanges later (i.e. "pay me or I invalidate the last year's worth of transactions").  The payout for the attacker would likely increase with time, so it's worth it for the attacker to be patient.

Per SIP-007, when a set of miners doesn't receive the anchor block, it recovers by burning BTC instead of transferring it -- these miners build one or more "nephew forks" that burn BTC during this reward cycle, since they descend from an uncle of the anchor block.  What we do _not_ want to have happen is for the nephew fork(s) produced by these nodes to get invalidated by a later-arriving anchor block _if_ they are sufficiently supported by the network.  That is, the network ought to continue to treat these nephew forks as valid **if more BTC was burned for nephew forks than was transferred for forks descending from the hidden anchor block**.  This way, a PoX-descendant fork set that descends from a hidden anchor block will only be considered as valid if it _also_ has majority miner support -- i.e. it represents more cumulative BTC transferred than the nephew fork set represents cumulative BTC burned.  This additional constraint makes pulling off a hidden anchor block reorg attack _at least_ as expensive as a 51% attack -- if a hidden anchor block attack plus a deep reorg attack did take place, then attacker would already have the power to reorg the chain however (s)he saw fit anyway.

We already store enough information in the sortition DB to do this.  We would need to make the following changes:

* When sending a block-commit transaction, the miner sets a bit in the `OP_RETURN` to indicate whether or not it's building off of a PoX-descendant fork, or building off of a nephew fork.  We have exactly eight bits of unused space in the block-commit wire format, so this is imminently doable.
* When processing sortition _S_ in reward cycle _N_, a node considers how much BTC was burned in all nephew forks in this reward cycle (i.e. sum of all BTC burned on nephew block-commits in _N_ up to and including _S_) versus how much BTC was transferred in all PoX-descendant forks (i.e. sum of all BTC transferred in PoX-descendant block-commits in _N_ up to and including _S_).  If more BTC has been burned than transferred, it only considers valid the sortitions from the nephew block-commits up to and including _S_.  If more BTC has been transferred, it only considers valid the sortitions from just the PoX-descendant block-commits.
* Once reward cycle _N+1_ begins, the miners will have decided whether or not reward cycle _N_ has a valid anchor block.  If the nephew fork set represents more BTC burnt than the PoX-descendant fork set represents BTC transferred, then it doesn't matter if the anchor block arrives -- it will be considered invalid, and no forks may be built on it.  The PoX bit for this reward cycle is set to `1`, and no anchor block is selected.  If on the other hand more BTC has been transferred, then the nodes decide that there was an anchor block in reward cycle _N_ and all nephew forks built on during _N_ are invalidated.  The PoX bit in reward cycle _N_ is set to `1` for nodes that have the anchor block, and `0` for nodes that do not.  The cumulative weight measurement for nephew forks versus PoX-descendant forks resets at the start of _N + 1_.

Point (3) is required because it means that the only way a malicious miner who mines hidden anchor blocks can keep stalling the chain is to _also_ transfer more BTC in a descendant PoX fork than all nephew forks combined during reward cycle _N_.  This is tantamount to forcing this miner to mount a 51% attack.  Moreover, if the attacker loses and a nephew fork out-burns it during reward cycle _N_, then the attacker must re-start the attack at reward cycle _N+1_ since its anchor block will not be considered by any node even if it gets released.

Per point (2), note that this means that if the PoX-descendant fork-set and the nephew fork-set have roughly equal mining power, then the chain will get reorged intermittently back to the start of _N_.  I consider this unlikely -- network participants can, through user-burn-supports, quickly step in and help honest miners overcome a hidden anchor block attacker.  Also, because all sortitions are public knowledge, every network participant -- including wallets and exchanges -- can see this attack happening, and take proactive security measures like increasing the number of required confirmations or temporarily halting trading.  So, the attacker would not have much to gain from triggering these intermittent re-orgs.

Due to point (3), the ability to re-org a reward cycle's forks goes away if honest miners are able to produce a nephew fork-set with more BTC burnt (in which case, the best chain is the longest chain in the nephew fork-set).  Even if an attacker is able to repeatedly mine hidden anchor blocks for reward cycles _N_ up to _N + k_, the remaining miners can recover once the attack subsides by building a fork that descends from the best chain tip at the start of _N_ in reward cycle _N + k + 1_.  Note that all miners and nodes that _don't_ have the anchor block for _N_ will have always considered this chain tip to be the best chain tip for the past _k + 1_ cycles, since they will not have been able to process any PoX-descendant forks.  If the attacker releases the anchor block at all, then it can only reorg the chain up to _k + 1_ reward cycles (but at the cost of having to constantly out-transfer all nephew-miners).

**Comment by jcnelson** [comment]
(oops, fat-fingered the "comment" button).

Anyway, the last thing I'll say is that another advantage of trying to make the network commit to a nephew fork versus a PoX-descendant fork by measuring burn/transfer output would be a far less disruptive change than trying to make the MARF block commitments _not_ depend on a PoX fork.  I also think that the implementation would be somewhat straightforward and wouldn't require any new data structures to be invented.

**Comment by kantai** [comment]
> 2. When processing sortition S in reward cycle N, a node considers how much BTC was burned in all nephew forks in this reward cycle (i.e. sum of all BTC burned on nephew block-commits in N up to and including S) versus how much BTC was transferred in all PoX-descendant forks (i.e. sum of all BTC transferred in PoX-descendant block-commits in N up to and including S). If more BTC has been burned than transferred, it only considers valid the sortitions from the nephew block-commits up to and including S. If more BTC has been transferred, it only considers valid the sortitions from just the PoX-descendant block-commits.


> 3. Once reward cycle N+1 begins, the miners will have decided whether or not reward cycle N has a valid anchor block. If the nephew fork set represents more BTC burnt than the PoX-descendant fork set represents BTC transferred, then it doesn't matter if the anchor block arrives -- it will be considered invalid, and no forks may be built on it. The PoX bit for this reward cycle is set to 1, and no anchor block is selected. If on the other hand more BTC has been transferred, then the nodes decide that there was an anchor block in reward cycle N and all nephew forks built on during N are invalidated. The PoX bit in reward cycle N is set to 1 for nodes that have the anchor block, and 0 for nodes that do not. The cumulative weight measurement for nephew forks versus PoX-descendant forks resets at the start of N + 1.


I agree that (2) makes this attack more difficult/unlikely, because it would mean that a malicious miner would need to secretly mine a PoX anchor block (non-trivial, need to win a super-majority of blocks during the prepare phase), and then _also_ provide a majority of the mining power during the subsequent reward cycle. However, I don't think that (3) addresses the "long-range attack" problem for the following case:

1. Malicious miner gets a PoX anchor block A confirmed for reward cycle N which is hidden to the honest network. Requires winning a super-majority of blocks during Prepare.
2. Malicious miner mines a majority of commits during the corresponding reward cycle N.

Honest miners would like to disqualify A from consideration after reward cycle N has passed, but they cannot. The network is now in as tenuous of a position as before: honest miners would like to continue making progress, but revealing A could disqualify their fork in all cases.

**Comment by kantai** [comment]
> If malicious miner M can produce private anchor block A, and consistently control a majority of the mining power such that the best known fork is M's private fork built on A, then there's not much the rest of the network can do. This is true even without PoX. An analogous attack on Bitcoin would be for M to produce the heaviest Bitcoin fork, but only release the block headers (thereby proving that the fork exists, but without allowing anyone to build on it except for empty blocks).

The point is that the miner M only needs to do this for a window of time to permanently disrupt the chain -- they need to win (and hide) an anchor block A, and then have the majority of burns in that block's associated reward cycle. After doing so, without ever revealing the anchor block, the chain is permanently in a state that it could be forked simply by revealing the block A.

**Comment by jcnelson** [comment]
> However, I don't think that (3) addresses the "long-range attack" problem for the following case:

I'm not sure I communicated this well enough in my point (2).  Specifically, I'm saying that nodes don't consider a sortition's BTC burned/xfered in a fork's cumulative BTC weight _until they have the Stacks block for it_.  So, if M mines a private fork, the rest of the network still considers the nephew fork to be the best fork, since it's the best fork _for which they have data_.  Another node won't count M's sortitions towards M's fork until it has M's fork's block data.

That said, now suppose malicious miner M produces a private anchor block A in reward cycle N, and then produces the heaviest fork in reward cycle N.  This fork is private for now -- only M sees the block contents, and it descends from A.

Suppose that M ceases the attack in reward cycle N+1.  Then, the rest of the network spends reward cycle N+1 building a nephew fork "in parallel" to M's private fork.  Then, one of the following can happen:

* M can release A before the nephew fork becomes heavier than M's private fork.  Per the rules in (2), this _by itself_ does not invalidate the nephew fork, since the nephew fork still has a heavier weight than M's fork.  This is because M has only disclosed A, and not the fork descending from A.  Therefore, per (2), it looks to the rest of the network that M's fork is a minority fork that stops at A -- M's subsequent sortitions don't yet count towards M's fork, because M hasn't released the corresponding block data.  I think this addresses the "long-range attack" problem -- M can't invalidate the nephew fork just by releasing A.

* M can release both A and its private fork before the nephew fork can become heavier.  Because M's fork is heavier than the nephew fork, all other miners switch to building off of it.  This is tantamount to a deep reorg -- M has reorged the chain by producing and then releasing a heavier fork.  Other nodes recognize M's fork as heavier once they receive and process all of M's block data.

* M releases A and its private fork after the nephew fork gets heavier.  The nodes receive and process the fork, but no reorg happens since the nephew fork has more cumulative BTC behind it.

I think this version of (2) means that in all scenarios, in order for M to reorg the chain, it needs to consistently control the majority of the mining power.  This is what we want -- M can't reorg the chain unless M has majority mining power.

**Comment by kantai** [comment]
Ah -- I think I see.

So in this case, when we consider whether or not to reorg a "PoX fork", we consider the burn weight of the two PoX forks. I think this would mean that we cannot invalidate PoX forks as we currently do-- a PoX fork may be invalid when considered at block height N, but it is valid at block height N+k (because it has overtaken the other fork in burn weight).

**Comment by jcnelson** [comment]
The fundamental problem here is that there exists a way for an attacker to do _O(1)_ work (i.e. to produce a private anchor block), and in doing so, can force the network to do _Theta(n)_ work later (i.e. by releasing it).  What we want is a mining protocol whereby a would-be attacker must _always_ do _Theta(n)_ work in order to compel the network to do _Theta(n)_ work (or in other words, the attacker must do _O(1)_ work per block in the attacker fork -- each block the network processes is paid for by sufficient mining power).

Without PoX, the Stacks blockchain already has this property.  Like all forkable blockchains, an attacker that wants to force the network to re-process _Theta(n)_ blocks must first mine a private fork of _Theta(n)_ blocks.  But with PoX, the attacker can mine a single hidden anchor block and release it arbitrarily far into the future, triggering a deep reorg of _Theta(n)_ blocks.

How can we stop this from happening in PoX?  We stop a similar problem in the Stacks blockchain by permitting forks:  just because a block is missing and released late doesn't cause the subsequent chain to be reorged.  Instead, miners who don't have a missing block simply mine on a different fork.  The attacker can mine its own fork if it wants, but it will have to do _Theta(n)_ work to produce a fork of _Theta(n)_ blocks.  This means that the only way a late block can later become part of the canonical chain history is if the _majority_ of the network mines a fork that includes it.

In PoX, in order to stop a late anchor block from disrupting the Stacks chain in a similar way, the Stacks chain needs to maintain a fork history of _anchor block decisions_.  An _anchor block decision_ is a collective decision made by miners during a reward cycle to classify an anchor block as being in one of two states:  the anchor block was either _known_ to the majority of miners during the reward cycle, or the anchor block was _unknown_ to the majority of miners during the reward cycle (this obviously does not apply to reward cycles without anchor blocks).  Miners affirm past anchor block decisions by attaching the current reward cycle to a _parent_ reward cycle, and in doing so, affirms the decisions made in all ancestral anchor blocks.

Reward cycle forks now represent a history of decisions made on the states of each ancestor's anchor block.  The canonical reward cycle chain is the longest reward cycle chain, and the canonical Stacks chain is the longest Stacks chain that resides on the canonical reward cycle fork.

Breaking this down, I think the system needs to track _three_ sets of forks:

* The set of burnchain forks.  The best burnchain fork is the longest one we know about (we let Bitcoin take care of feeding us the right blocks).
* The set of reward cycle forks.  The best reward cycle fork is the longest chain of reward cycles on the best burnchain fork.
* The set of Stacks forks.  The best Stacks fork is the longest Stacks fork in the longest reward cycle fork.

## Reward Cycle Forks

Reward cycles and prepare phases would continue to operate at fixed intervals of burnchain blocks in which PoX (or PoB) happen, just as they do today.  But now, miners will create a fork history of reward cycles.  The _Nth_ reward cycle is no longer required to descend from reward cycle _N - 1_, but may descend from any reward cycle from 0 up to _N - 1_.  Through a majority vote, miners in reward cycle _N_ decide which reward cycle will be its parent (see below).  This decision-making already happens implicitly; we just need to track it so we can enforce the best-Stacks-fork rule above.

Reward cycle forks are central to preventing a late anchor block from triggering more than _O(1)_ processing work.  A reward cycle's sortitions and blocks will _never_ be re-processed if an anchor block was chosen but miners decided it did not arrive in time.  But at the same time, the network must be able to recover if the majority of the mining power decides that an anchor block _did_ arrive, even though it remains private (or gets lost).  Creating a fork history of the anchor block decisions achieves this.

Because the same burnchain blocks can encode multiple reward cycles on different reward cycle forks, it is important to only process an anchor block if the majority of the network _currently_ believes it exists.  This is necessary in order to prevent an attacker from being able to get nodes to re-process sortitions on minority reward cycle forks.  To achieve this, nodes would need to adhere to the following strategies when considering an anchor block:

* If the anchor block arrives for a non-canonical reward cycle fork, it is ignored.  This is trivial -- not only are anchor blocks for non-canonical reward cycle forks ignored, but _all_ blocks from them are ignored, because the node will not have processed their sortitions.  In other words, an anchor block will only be processed at all if its reward cycle is canonical.
* If a late anchor block arrives for a reward cycle on the canonical reward cycle fork, it is only processed if the majority of mining power had the anchor block at the time the reward cycle was mined.  Otherwise, it is ignored.  The sortitions in the reward cycle encode the decision the miners made at the time.

Breaking this down, we can implement these two strategies by doing the following:

* Make it so when mining blocks in reward cycle _N_, miners vote on which reward cycle it its parent.  When a miner produces a block, the highest ancestor block that is not in _N_ will be in some other reward cycle _N - k_.  If the majority of blocks in _N_ pick these ancestors in the same _N - k_, then the parent reward cycle is _N - k_.  If there is no majority consensus, then the parent reward cycle is 0.

* Make it so that if reward cycle _N_ has an anchor block, but a miner does not have it, then the miner can signal the absence of the anchor block in a PoB block-commit.  Then, if the node later receives the anchor block (or if a bootstrapping node receives it), the node will first count up how many PoB sortitions were won during reward cycle _N_.  If the majority of them were PoB and signaled that the anchor block was absent, then the anchor block and all of its descendant Stacks blocks are ignored.

## Choosing the Parent Reward Cycle

To create a fork history of reward cycles, we need to determine which reward cycle is the parent reward cycle of reward cycle _N_.  To do so, the node examines each block mined reward cycle _N_, and determines its highest ancestor that is _not_ in reward cycle _N_ (i.e. via the `last_ancestor_of_block_before()` algorithm in SIP-007).  The reward cycle of this highest ancestor is some reward cycle _N - k_, and the act of mining a descendant of this block is a _vote_ for reward cycle _N - k_ to be the parent of reward cycle _N_.

Simply by mining blocks, miners vote on what the parent reward cycle should be.  If a majority of blocks in reward cycle _N_ all descend from highest ancestors in the same reward reward cycle _N - k_, then the parent of _N_ is _N - k_.  If no majority can be found, then reward cycle _N_'s parent is reward cycle 0.

## Deciding the Fate of a Late Anchor Block

In addition to voting on which reward cycle should be the parent of _N_, miners vote to confirm whether or not the network had the anchor block for _N_ while _N_ was being mined.  This already happens implicitly -- if a miner does not have the anchor block for _N_, it can still mine through PoB during _N_ by building a "nephew chain" that descends from an uncle of the anchor block.  In doing so, miners are implicitly casting votes on whether or not _N_'s anchor block existed at the time.

We can track this vote to prevent a late anchor block from forcing a deep chain reorg.  Recall that one of three things can happen when processing reward cycle _N_:

* No anchor block gets chosen (and PoB happens)
* An anchor block gets chosen, and the miner receives and processes it (and PoX happens)
* An anchor block gets chosen, but the miner does not receive it (in which case, only PoB block-commits will be processed)

If the reward cycle has an anchor block chosen, miners have to decide during the reward cycle whether or not it falls under cases (2) or (3) above.  If it falls under case (3), then miners send PoB block-commits with an additional bit set in them to indicate that the reason the miner is sending a PoB block-commit is because it doesn't have this reward cycle's anchor block.  Then, when a node begins to process the sortitions this reward cycle, it will see whether or not a majority or a minority of the winning sortitions are PoB sortitions with this "anchor-block-missing" bit set.  If it's a majority, then case (3) applies to this a reward cycle, and the node will never accept an anchor block for it.  If it's a minority, then case (2) applies and the node will accept an anchor block for it even if it arrives late, but only as long as this reward cycle is on the canonical reward cycle fork.

## Recovering from a Long 51% Attack

Both of the above strategies are required to handle the case where a malicious miner is able to successively create a private anchor block for reward cycles _N - k_ through _N_, and mine a majority of blocks in _N - k_ through _N_ through PoX sortitions.  If such a miner exists, the miner was necessarily a 51% attacker for reward cycles _N - k_ through _N_.  The two strategies proposed here give the honest majority of the network a chance to recover and orphan the attacker's chain once the attack stops.

To do so, the honest majority of the network uses reward cycle _N + 1_ to begin building a nephew Stacks chain that will overtake the attacker's chain.  The miners declare reward cycle _N + 1_ to be a child of reward cycle _N - k - 1_.  Then, miners declare that reward cycle _N + 2_ is a child of reward cycle _N + 1_, that _N + 3_ is a child of _N + 2_, etc. until the fork of reward cycles _0, 1, ..., N - k - 2, N - k - 1, N + 1, N + 2, ..., N + k + 1_ is one reward cycle longer than the attacker fork _0, 1, ..., N - k - 2, N - k, N - k - 1, ..., N - 1, N_.  At this point, the longest Stacks fork represented in reward cycle _N + k + 1_ becomes the canonical Stacks chain.

Once the reward cycle fork _N + k + 1_ overtakes the attacker fork at _N_, the attacker cannot force a deep reorg by releasing its private anchor blocks and descendants.  This is because the attacker's fork now resides on a non-canonical reward cycle fork.  However, the attacker is able to force deep reorgs back to the anchor block of _N - k_ by releasing anchor blocks and their descendants while the network is recovering.  But, the attacker can only reorg reward cycles in which it had a majority of the mining power, which means that instigating a deep reorg of _Theta(n)_ blocks requires the attacker to do _Theta(n)_ work.

**Comment by kantai** [comment]
I think this proposed plan sounds like a good one! With this in place, we don't need to ensure that MARF block commitments don't depend on PoX forks.

**Comment by jcnelson** [comment]
> Make it so that if reward cycle N has an anchor block, but a miner does not have it, then the miner can signal the absence of the anchor block in a PoB block-commit. Then, if the node later receives the anchor block (or if a bootstrapping node receives it), the node will first count up how many PoB sortitions were won during reward cycle N. If the majority of them were PoB and signaled that the anchor block was absent, then the anchor block and all of its descendant Stacks blocks are ignored.

Hidden in this requirement is a necessary change to how we process sortitions:  If the node has a reward cycle's anchor block and is processing PoX block-commits in a given block, then it must also consider all PoB sortitions in the same block that vote to reject the anchor block (but ignore PoB sortitions that do not vote to reject).  Also, if the node does _not_ have a reward cycle's anchor block, but the reward cycle chose one, then each sortition must consider _all_ PoB sortitions, _regardless_ if they vote to reject the anchor block or not.

The reason this is needed is because when a reward cycle is being mined, some miners will have the anchor block and some will not.  If a bootstrapping node has the anchor block for this reward cycle, then it will be able to validate PoX sortitions even if some miners were unable to at the time.  But in order to decide whether or not to ignore the anchor block even though it has it, the protocol must be able to permit it to consider block-commits that reject the anchor block so they'll know how to treat this anchor block.  This means that the block-commits eligible for sortition must be both the PoX block commits (which are validated against the anchor block) and the PoB/reject block-commits.  If on the other hand the bootstrapping node does not have the anchor block even though one was chosen, it must still process both PoB/accept and PoB/reject sortitions, so it can know what to do with an anchor block should it ever arrive later and become part of the canonical chain state.

**Comment by jcnelson** [comment]
Had a call with @kantai about this.  Upon further reflection, this scheme cannot work by itself -- there's no way for a bootstrapping node to know in advance whether or not a reward cycle is going to end up on the canonical reward-cycle fork.  This is because in order to determine if a reward cycle is canonical, it needs to know the length of all descendant reward cycle forks -- it cannot calculate this incrementally, because the correct interpretation of this reward cycle's sortitions may require interpreting the sortitions made by the network _minority_ (but which will later be confirmed by the network _majority_).  So in order to calculate this, the node needs to have some "future knowledge" -- namely, it needs to know exactly which future anchor blocks will be rejected or accepted as a consequence of accepting/rejecting this current anchor block.  There's no reliable way to propagate this information with just the above proposal.

An alternative approach proposed by @kantai would be to require that the prepare phase of each reward cycle always be proof-of-burn, and require that the prepare phase and _only_ the prepare phase sortitions vote on whether or not to accept or reject the current anchor block, and which reward cycle shall be the parent reward cycle of the current reward cycle.  This would make it so that a bootstrapping node would always be able to extract the accept/reject history and fork history of all reward cycles regardless of which Stacks blocks it already has.

However, implementing and testing the above would likely take more time than we have (two or more weeks at least), so this issue might need to be addressed in Stacks 2.1.  A simpler strategy for Stacks 2.0 could be to simply implement an on-chain permanent-kill switch for PoX, which liquid STX holders activate in the unlikely-but-catastrophic event that a PoX anchor block goes missing for too long.  We could remove the kill switch in 2.1 when this fork-recover mechanism is ready.

**Comment by jcnelson** [comment]
I spent some time working through the solution space a bit more.  I think the constraints we're working with are as follows:

* Whatever we do, this is going to be an on-Bitcoin protocol.  Participants must communicate through Bitcoin transactions.
* However we do this, the on-Bitcoin transactions must be _self-evidently valid_.  That is, a node must be able to decide whether or not an on-Bitcoin transaction is valid without having any knowledge of the Stacks chain state.  **This precludes a protocol that involves measuring STX participation**, since a node will only be able to validate a STX-holder vote if they can first determine that the represented address does indeed possess that many STX.

Point (2) means that if there is some kind of on-Bitcoin voting procedure to reject the anchor block, then the voting _must_ be done through PoB, not PoX.  The vote must be conducted during some pre-determined, dedicated block-races that are PoB-only.

We aren't limited to using a miner-driven voting procedure to handle missing anchor blocks.  Yet another alternative solution space is to implement a majority-vote among a set of trusted keys, for example, which are entrusted to a set of principals whose sole responsibility is to identify missing anchor blocks.  These key-holders can overrule miners and declare an anchor block missing.  I'm leaving out a lot of important details on how these keys get distributed and how to choose key-holders, but I bring this up as a possible solution that doesn't involve significantly altering how PoX works today.

**Comment by jcnelson** [comment]
Wrapping up notes from the architecture meeting, the consensus going forward is to attack this problem in two phases:

1.  Before code-freeze, make it so that the prepare phase is 100 blocks, and is always PoB.  If we can do this, then we can implement the rest of this issue as a soft fork.  Will open a separate issue for this.
2.  Once (1) is done, have the node's sortition DB keep track of a fork history of reward cycles, and use that to determine which anchor blocks to unconditionally ignore (i.e. non-canonical reward cycles are always treated as PoB).  It would be wonderful, but not required, for this to be ready by January 14.

**Comment by muneeb-ali** [comment]
It was great to discuss this live in today's call. I'm supportive of #2157 with proof-of-burn only prepare phase and trying to make the prepare phase shorter.

**Comment by jcnelson** [comment]
Wrote up a possible solution to this problem here: https://gist.github.com/jcnelson/b1aa4bef8b9adb0856b28d3a933ef9a0

**Comment by jcnelson** [comment]
Merged!
